packages/web3-common/src/errors.ts

```diff
@@ -0,0 +1,333 @@
+/* eslint-disable max-classes-per-file */
+
+import {
+	ERR_RESPONSE,
+	ERR_PARAM,
+	ERR_CONN,
+	ERR_CONN_INVALID,
+	ERR_CONN_TIMEOUT,
+	ERR_CONN_NOT_OPEN,
+	ERR_CONN_CLOSE,
+	ERR_CONN_MAX_ATTEMPTS,
+	ERR_CONN_PENDING_REQUESTS,
+	ERR_INVALID_PROVIDER,
+	ERR_INVALID_RESPONSE,
+	ERR_TX_REVERT_INSTRUCTION,
+	ERR_TX_REVERT_TRANSACTION,
+	ERR_TX,
+	ERR_TX_NO_CONTRACT_ADDRESS,
+	ERR_TX_CONTRACT_NOT_STORED,
+	ERR_TX_REVERT_WITHOUT_REASON,
+	ERR_TX_OUT_OF_GAS,
+	ERR_CONTRACT_RESOLVER_MISSING,
+	ERR_CONTRACT_ABI_MISSING,
+	ERR_CONTRACT_REQUIRED_CALLBACK,
+	ERR_CONTRACT_EVENT_NOT_EXISTS,
+	ERR_CONTRACT_RESERVED_EVENT,
+	ERR_CONTRACT_MISSING_DEPLOY_DATA,
+	ERR_CONTRACT_MISSING_ADDRESS,
+	ERR_CONTRACT_MISSING_FROM_ADDRESS,
+} from './constants';
+
+type ConnectionEvent = { code: string; reason: string };
+type Response<T = unknown> = { error?: { message?: string; data?: T } };
+type Receipt = Record<string, unknown>;
+
+export abstract class Web3Error extends Error {
+	public readonly name: string;
+	public readonly abstract code: number;
+
+	public constructor(msg: string) {
+		super(msg);
+		this.name = this.constructor.name;
+		Error.captureStackTrace(this, Web3Error);
+	}
+
+	public toJSON() {
+		return { name: this.name, code: this.code, message: this.message };
+	}
+}
+
+export class InvalidNumberOfParamsError extends Web3Error {
+	public code = ERR_PARAM;
+
+	public constructor(public got: number, public expected: number, public method: string) {
```

Comment:
I thought constructors were inherently `public`?

Reply:
Yes the constructors can also be `protected` and in TS `private` as well. Where only class own's static members are allowed to call those private constructors. 

To improve the readability of the code our Eslint rules enforce to must specify the visibility..